homeassistant/components/xiaomi_vacuum.py

```diff
@@ -0,0 +1,418 @@
+        return self._icon
+
+    @property
+    def available(self):
+        """Return true when state is known."""
+        return self._handler.state is not None
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the device."""
+        return self._handler.state_attrs
+
+    @property
+    def is_on(self):
+        """Return true if switch is on."""
+        return self._handler.is_on
+
+    @asyncio.coroutine
+    def async_turn_on(self, **kwargs):
+        """Turn the vacuum on."""
+        yield from self.hass.async_add_job(self._handler.turn_on_cleaning)
+
+    @asyncio.coroutine
+    def async_turn_off(self, **kwargs):
+        """Turn the vacuum off and return to home."""
+        yield from self.hass.async_add_job(self._handler.turn_off_cleaning)
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Fetch state from the device."""
+        yield from self.hass.async_add_job(self._handler.update_vacuum_state)
+
+
+class MiroboVacuumSensor(Entity):
+    """Representation of a sensor of a Xiaomi Vacuum cleaner."""
+
+    def __init__(self, mirobo_vacuum, name, sensor_type):
+        """Initialize the sensor object."""
+        self._handler = mirobo_vacuum
+        self._sensor = sensor_type
+
+        friendly_name, unit, icon = SENSOR_MAP[sensor_type]
+        self._name = '{}_{}'.format(name, sensor_type)
+        self._friendly_name = friendly_name
+        self._icon = icon
+        self._unit = unit
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit the value is expressed in."""
+        return self._unit
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return getattr(self._handler.state, self._sensor)
+
+    @property
+    def should_poll(self):
+        """Return True if entity has to be polled for state."""
+        return False
+
+    @property
+    def available(self):
+        """Return true when state is known."""
+        return self._handler.state is not None
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        attrs = {"friendly_name": self._friendly_name}
+        return attrs
+
+    @property
+    def icon(self):
+        """Return the icon for the sensor."""
+        if self._sensor == 'battery':
+            returning_icon = 'mdi:battery'
+            if not self.available:
+                return returning_icon + '-unknown'
+            if getattr(self._handler.state, 'state') == 'Charging':
+                returning_icon += '-charging'
+            if 20 < self.state < 100:
+                returning_icon += '-{}'.format(int(self.state / 20) * 20)
+            elif self.state < 20:
+                returning_icon += '-outline'
+            return returning_icon
+        return self._icon
```

Comment:
well, it seems like the sensors aren't very orthodox, so, we'll see how this comes out...